### PR TITLE
Add `s.merge` to merge mapping Specs

### LINKED ISSUE #70: Add functionality for merging mapping Specs
It is not possible to merge multiple mapping Specs together right now without defining a custom Spec. Using `s.all` would result in losing any keys whose Specs are defined in later Specs during conformation.

### PR DESCRIPTION
**Pull request by chrisrink10**
Adds a new Spec type to merge mapping Specs (under `s.merge`).

As part of this change, I moved `all_spec` and `any_spec` into `base.py`, because they were needed in the merge spec factory body, but I think they ended up making more sense in `base` since combination Specs are kind of a base-level feature of `dataspec`. I also broke out `ObjectSpec` from being a subclass of `DictSpec`, since `ObjectSpec` keys _must_ be strings whereas `DictSpec` simply accepts any hashable object as a key.

I ended up discovering that it was possible to duplicate keys between required and optional keys in mapping spec definitions (since optional keys are wrapped in `s.opt` and thus do not compare hash-equal), so I fixed that bug too.

Fixes #70, #74

### CODE REVIEW

File: tests/test_base.py [hunk 1]
```diff
@@ -248,6 +249,13 @@ def test_set_coll_conformation(self, set_spec: Spec):
 
 
 class TestDictSpecValidation:
+    @pytest.mark.parametrize(
+        "pred", [{"id": str, s.opt("id"): int}, {s.opt("id"): int, "id": str},]
+    )
+    def test_dict_spec_definition(self, pred):
```
Comment (Palisand) [inline tests/test_base.py:255]: Can you add a comment on the fact that this is testing duplicate keys please? Or maybe just change the test name.

File: tests/test_base.py [hunk 2]
```diff
@@ -686,6 +701,292 @@ def test_namedtuple_conformation(self, namedtuple_spec: Spec):
...
+            s.merge({"id": int}, s.str("name"))
+
+        with pytest.raises(TypeError):
+            s.merge("with_tag", s.num(min_=0), {"name": str})
+
+    def test_no_overlapping_required_and_optional(self):
+        with pytest.raises(KeyError):
+            s.merge(
+                {"id": int, "name": str}, {"id": lambda v: v > 0, s.opt("name"): str}
+            )
```
Comment (Palisand) [inline tests/test_base.py:930]: Is it worth adding a test where only the required or optional keys (but not both) overlap?

File: src/dataspec/base.py [hunk 3]
```diff
@@ -589,25 +615,70 @@ def validate(self, d) -> Iterator[ErrorDetails]:  # pylint: disable=arguments-di
...
+    ) -> Spec:
+        assert len(specs) >= 2, "Must merge at least two Specs"
+
+        map_pred: MutableMapping[Hashable, List[SpecPredicate]] = defaultdict(list)
+        for spec in specs:
+            for k, pred in spec._reqkeyspecs.items():
+                map_pred[k].append(pred)
+
+            for k, pred in spec._optkeyspecs.items():
+                map_pred[OptionalKey(k)].append(pred)
```
[the remark this answers is not in the document]
Comment (chrisrink10) [reply src/dataspec/base.py:634]: Fair call out. There are no places (to my knowledge) in dataspec which rely on the insertion order of keys for dicts, so I think this is ok.
